insights/parsers/libkeyutils.py

```diff
@@ -1,18 +1,115 @@
+verify linked libraries.
+
+https://www.welivesecurity.com/2017/10/30/windigo-ebury-update-2/
 """
+import re
+
 from .. import Parser, parser
 
 
 @parser("libkeyutils")
 class Libkeyutils(Parser):
+    """
+    This parser finds all 'libkeyutils.so*' libraries in either /lib or /lib64 directory and its
+    sub-directories.
+
+    Output of Command::
+
+        /lib/libkeyutils.so.1
+        /lib/tls/libkeyutils.so.1.6
+        /lib64/libkeyutils.so
+
+    Example::
+
+        >>> shared[Libkeyutils].libraries
+        ['/lib/libkeyutils.so.1', '/lib/tls/libkeyutils.so.1.6', '/lib64/libkeyutils.so']
+    """
+    def __init__(self, *args, **kwargs):
+        self.libraries = []
+        """list: all 'libkeyutils.so*' libraries located in either /lib or /lib64 directory and its sub-directories."""
+        super(Libkeyutils, self).__init__(*args, **kwargs)
+
     def parse_content(self, content):
-        self.data = list(content)
+        self.libraries = list(content)
 
 
 @parser("libkeyutils_objdumps")
 class LibkeyutilsObjdumps(Parser):
+    """
+    This parser goes through objdumps of all 'libkeyutils.so.1' libraries in either /lib or /lib64
+    directory, and its sub-directories, to finds linked libraries.
+
+    Output of Command::
+
+        /lib/libkeyutils.so.1:     file format elf32-i386
+        /lib/libkeyutils.so.1
+        architecture: i386, flags 0x00000150:
+        HAS_SYMS, DYNAMIC, D_PAGED
+        start address 0x00000f80
+        ...
+
+        Dynamic Section:
+          NEEDED               libdl.so.2
+          NEEDED               libc.so.6
+          NEEDED               libsbr.so
+          SONAME               libkeyutils.so.1
+          INIT                 0x00000e54
+        ...
+
+
+        /lib64/libkeyutils.so.1:     file format elf64-x86-64
+        /lib64/libkeyutils.so.1
+        architecture: i386:x86-64, flags 0x00000150:
+        HAS_SYMS, DYNAMIC, D_PAGED
+        start address 0x00000000000014b0
+        ...
+
+        Dynamic Section:
+          NEEDED               libdl.so.2
+          NEEDED               libsbr.so.6
+          NEEDED               libfake.so
+          SONAME               libkeyutils.so.1
+          INIT                 0x0000000000001390
+        ...
+
+    Example::
+
+        >>> shared[LibkeyutilsObjdumps].linked_libraries
+        {'/lib/libkeyutils.so.1': ['libdl.so.2', 'libc.so.6', 'libsbr.so'],
+         '/lib64/libkeyutils.so.1': ['libdl.so.2', 'libsbr.so.6', 'libfake.so']}
+    """
+    FILE_PATTERN = re.compile(r'(.*libkeyutils.so.1):\s*file format')
+    NEED_PATTERN = re.compile(r'NEEDED\s+(.*)\s*$')
+
+    def __init__(self, *args, **kwargs):
+        self.linked_libraries = {}
+        """dict: found libraries and their linked libraries."""
+        super(LibkeyutilsObjdumps, self).__init__(*args, **kwargs)
+
     def parse_content(self, content):
-        self.data = list(content)
+        file_name = None
```

Comment:
I think the `file_name = None` line should be moved one line down - to be at the beginning inside the `for` loop.

With the current code as it is, I *think* that if there is file 1 that matches only FILE_PATTERN, it sets `file_name`, and then if there is another file 2 that **doesn't** match FILE_PATTERN but does match NEED_PATTERN, it will get improperly recorded into `linked_libraries`. I didn't test it though.

Reply:
I'm not so sure but maybe some static binaries can have nothing in NEEDED.

Reply:
The parser only runs on object dumps from `libkeyutils.so.1` files. See the `specs.py`, the command parsed is `find -L /lib /lib64 -name libkeyutils.so.1 -exec objdump -x "{}" \;`.

Reply:
@skontar did you resolve @jsvob's question about whether or not `file_name = None` should be moved into the loop?

Reply:
@skontar I missed that - in this case, I think the code is fine as it is. @bfahr I deem it resolved and I approve of the request. Sorry for not responding in time.

Reply:
@jsvob - Thanks for the reply, I'm going to merge it.